### PR TITLE
implement trace sampling

### LINKED ISSUE #228: Trace sampling
related:
- #29 
- #50
- #193
- #237 
- #224
- https://github.com/tokio-rs/tracing/issues/1768

request tracing can be expensive, we should have a way to select which traces should be sent to jaeger or opentelemetry.

### PR DESCRIPTION
**Pull request by Geal**
follow up on #193
fix #228

* moves tracing configuration to the `apollo_router::trace` module

### CODE REVIEW

**Comment by Geal** [comment]
so I've tested sampling a bit more, and we end up with two different approaches, tested in https://github.com/apollographql/router/pull/224/commits/535e8ae9738bf1b9f2b13bb33da5156729eefa92:
- [opentelemetry's samplers](https://docs.rs/opentelemetry/latest/opentelemetry/sdk/trace/trait.ShouldSample.html): This is called once, at the root span's creation, so it has to decide right there if the entire trace should be recorded or not, and it does not get any info on this trace or previous ones
- [tracing's Layer](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/layer/trait.Layer.html): they can be called on every span, see all the recorded data, add more info, etc

For the sampling strategy, I wanted initially a sampler that can decide depending on the request (example: keep 1 in 10 successful queries, but send all queries for which there were errors), but that does not play well with distributed tracing. If we want to refuse or accept an entire trace, we need to buffer all the spans until we know if it will be sent, and that can have a huge memory overhead (this is what [refinery](https://github.com/honeycombio/refinery) is doing). That also means that subgraphs could send some spans that would not be linked to spans from the router (because the router can't decide on sampling until subgraphs have answered).

To sum up we end up with the same issues as described in the [honeycomb blogpost](https://www.honeycomb.io/blog/getting-at-the-good-stuff-how-to-sample-traces-in-honeycomb/)

**Comment by Geal** [comment]
it would still be possible to make a sampler that changes the rate dynamically by looking at the spans. Example: if we see more spans with errors, we increase the rate, if we see an increase in traffic, we decrease the rate

**Comment by Geal** [comment]
So in the end this is just a thin layer over the work in #237. Trace sampling could already work with OTLP, so I made sure it could work with Jaeger too (they use the same code under the hood for that). And I added some documentation on the configuration for tracing and sampling.
The configuration supports a "parent based" sampling alternative, but I decided to hide that one because it is not clear right now if it should be configurable here of it is internal to the router. The various options are here: https://docs.rs/opentelemetry/latest/opentelemetry/sdk/trace/enum.Sampler.html

**Review by o0Ignition0o — APPROVED** [review]
